### PR TITLE
add UnregisterProposal

### PR DESCRIPTION
**Pull request by antaveiv**
more unit tests in progress

### CODE REVIEW

File: cmd/commands/server/command_server.go [hunk 1]
```diff
@@ -132,10 +133,24 @@ func (cmd *Command) Wait() error {
...
-	err := cmd.dialogWaiter.Stop()
+
+	providerId, err := cmd.identityLoader()
+	if err != nil {
+		return err
+	}
+
+	signer := cmd.createSigner(providerId)
+
+	err = cmd.mysteriumClient.UnregisterProposal(providerId.Address, signer)
```
Comment (tadovas) [inline cmd/commands/server/command_server.go]: You don't need to do anything in this place. Kill command signals openvpn process to exit, openvpn process reports exiting state, and then pinger is stopped and proposal unregistered.

File: cmd/commands/server/command_server.go [hunk 2]
```diff
@@ -93,15 +95,14 @@ func (cmd *Command) Start() (err error) {
 		case openvpn.ExitingState:
 			log.Info("Open vpn service exiting")
 			close(stopPinger)
+			cmd.mysteriumClient.UnregisterProposal(providerID.Address, signer)
```
Comment (tadovas) [inline cmd/commands/server/command_server.go]: Don't swallow possibly returned error - at least log it.

File: server/dto/node_unregister_request.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package dto
+
+type NodeUnregisterRequest struct {
```
Comment (tadovas) [inline server/dto/node_unregister_request.go]: ProposalUnregisterRequest ?

Comment (Waldz) [reply server/dto/node_unregister_request.go]: `NodeUnregisterRequest` is ok also, it's more consistent with remote endpoint naming

File: server/interface.go [hunk 4]
```diff
@@ -12,6 +12,7 @@ type Client interface {
 
 	FindProposals(providerID string) (proposals []dto_discovery.ServiceProposal, err error)
 	RegisterProposal(proposal dto_discovery.ServiceProposal, signer identity.Signer) (err error)
+	UnregisterProposal(providerId string, signer identity.Signer) (err error)
```
Comment (Waldz) [inline server/interface.go]: Be consistent with another method in interface `PingProposal(proposal)`:
- pass full `proposal` structure
- real adapted takes that he needs for unregistration DTO

File: cmd/commands/server/command_server.go (hunk 2 above)
Comment (Waldz) [reply cmd/commands/server/command_server.go]: Also this unregistration+logging is duplicated. I would like same cleanup function here e.g. `cmd.Kill()`

File: cmd/commands/server/command_server.go [hunk 5]
```diff
@@ -132,10 +133,24 @@ func (cmd *Command) Wait() error {
 // Kill stops server
 func (cmd *Command) Kill() error {
 	cmd.vpnServer.Stop()
-	err := cmd.dialogWaiter.Stop()
+
+	providerId, err := cmd.identityLoader()
```
Comment (Waldz) [inline cmd/commands/server/command_server.go]: Identity is loaded in startup, loading 2nd time looks too excess

File: cmd/commands/server/command_server.go (hunk 1 above)
Comment (Waldz) [reply cmd/commands/server/command_server.go]: Should we depend unregistration after openvpn exit? I think unregistration can be ran separately even before openvpn exit

File: server/mysterium_api_fake.go [hunk 6]
```diff
@@ -29,6 +29,28 @@ func (client *ClientFake) RegisterProposal(proposal dto_discovery.ServiceProposa
 	return nil
 }
 
+//UnregisterProposal unregisters a service proposal when client disconnects
+func (client *ClientFake) UnregisterProposal(providerID string, signer identity.Signer) error {
+	matchingProposals, err := client.FindProposals(providerID)
```
Comment (Waldz) [inline server/mysterium_api_fake.go]: I suggest not to couple on function `FindProposals`, just traverse `client.proposalsMock` and delete matched values

File: cmd/commands/server/command_server.go (hunk 1 above)
Comment (Waldz) [reply cmd/commands/server/command_server.go]: But mostly I miss tests which imitates all those cases we talk here

Comment (tadovas) [reply cmd/commands/server/command_server.go]: Actually we should call unregister AFTER pinger is stopped. So basically the best place to put Unregister is on command_server.go:117, where stopPinger channel case is handled.

Comment (Waldz) [reply cmd/commands/server/command_server.go]: So it is:
```
stop service (vpnServer, dialogWaiter, natService) -> stop pinger -> unregister proposal
```

File: server/dto/node_unregister_request.go (hunk 3 above)
Comment (tadovas) [reply server/dto/node_unregister_request.go]: Fine by me , until this structure not leaked anywhere else out of mysterium api scope

File: cmd/commands/server/command_server_test.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	myID                  = identity.FromAddress("identity-1")
+	activeProviderID      = identity.FromAddress("vpn-node-1")
+	activeProviderContact = dto_discovery.Contact{}
+	activeProposal        = dto_discovery.ServiceProposal{
+		ProviderID:       activeProviderID.Address,
+		ProviderContacts: []dto_discovery.Contact{activeProviderContact},
+	}
+)
+
+func (tc *testContext) TestProposalUnregisteredWhenPingerClosed() {
```
Comment (tadovas) [inline cmd/commands/server/command_server_test.go]: Simple TestXXX(t * testing.T) is enough here.
testContext structure is needed when you want to use Setup/Teardown callbacks before each test or test suite.

File: cmd/commands/server/command_server_test.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"github.com/stretchr/testify/suite"
+	"github.com/mysterium/node/server"
+	"github.com/mysterium/node/identity"
+	dto_discovery "github.com/mysterium/node/service_discovery/dto"
+	"github.com/stretchr/testify/assert"
+)
+
+type testContext struct {
```
Comment (tadovas) [inline cmd/commands/server/command_server_test.go]: This structure is not needed

File: cmd/commands/server/command_server.go [hunk 9]
```diff
@@ -132,11 +119,34 @@ func (cmd *Command) Wait() error {
...
 		return err
 	}
+
 	err = cmd.natService.Stop()
 
 	return err
 }
+
+// Ping proposal until stopped, then unregister proposal
+func PingProposalLoop(proposal dto_discovery.ServiceProposal, mysteriumClient  server.Client, signer identity.Signer, stopPinger <- chan int) {
```
Comment (tadovas) [inline cmd/commands/server/command_server.go]: This function is used inside package only - it can be private

Comment (Waldz) [inline cmd/commands/server/command_server.go]: It's internal function, no point to make it public

File: cmd/commands/server/command_server_test.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+	fakeDiscoveryClient.RegisterProposal(activeProposal, nil)
+
+	go PingProposalLoop(activeProposal, fakeDiscoveryClient, nil, stopPinger)
+
+	close(stopPinger) //causes proposal to be unregistered
+
+	proposals, err := fakeDiscoveryClient.FindProposals(activeProposal.ProviderID)
+
+	assert.NoError(tc.T(), err)
+	assert.Empty(tc.T(), proposals)
```
Comment (Waldz) [inline cmd/commands/server/command_server_test.go]: `Empty` checks if object is empty, I would rather check length of proposals array

File: cmd/commands/server/command_server_test.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+type testContext struct {
+	suite.Suite
+}
+
+var (
+	myID                  = identity.FromAddress("identity-1")
+	activeProviderID      = identity.FromAddress("vpn-node-1")
+	activeProviderContact = dto_discovery.Contact{}
+	activeProposal        = dto_discovery.ServiceProposal{
```
Comment (Waldz) [inline cmd/commands/server/command_server_test.go]: You dont need so rich proposal, having `dto_discovery.ServiceProposal{}` would be enought

**Comment by tadovas** [comment]
Closed as irrelevant

**Comment by Waldz** [comment]
Comments fixed in #232

**Pull request closed without merging**